worker/uniter/resolver/loop.go

```diff
@@ -167,3 +176,53 @@ func updateCharmDir(logger Logger, opState operation.State, guard fortress.Guard
 		return guard.Lockdown(abort)
 	}
 }
+
+func checkCharmUpgrade(logger Logger, remote remotestate.Snapshot, rf *resolverOpFactory, ex operation.Executor) error {
+	// If we restarted due to error with a pending charm upgrade available,
+	// do the upgrade now.  There are cases (lp:1895040) where the error was
+	// caused because not all units were upgraded before relation-created
+	// hooks were attempted for peer relations.  Do this before the remote
+	// state watcher is started.  It will not trigger an upgrade, until the
+	// next applicationChanged event.  Could get stuck in an error loop.
+
+	local := rf.LocalState
+	local.State = ex.State()
+
+	// If the unit isn't started or already upgrading, no need to start an upgrade.
+	if !local.State.Started || local.State.Kind == operation.Upgrade ||
+		(local.State.Hook != nil && local.State.Hook.Kind == hooks.UpgradeCharm) ||
+		remote.CharmURL == nil {
+		return nil
+	}
+
+	if *local.CharmURL == *remote.CharmURL {
+		return nil
+	}
+
+	if remote.CharmProfileRequired {
+		if remote.LXDProfileName == "" {
+			return nil
+		}
+		rev, err := lxdprofile.ProfileRevision(remote.LXDProfileName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if rev != remote.CharmURL.Revision {
+			logger.Tracef("Charm profile required: current revision %d does not match new revision %d", rev, remote.CharmURL.Revision)
```

Comment:
Should this be trace?

Reply:
It's the same as in verifyCharmProfileResolver.  Perhaps both should be Debug?

Reply:
Yeah, seems like it would be valuable to know that a charm profile is required without going to a trace logging.